File: dotnet/SeleniumCore/By.cs
using System;

namespace OpenQA.Selenium;

/// <summary>
/// A locator: a (strategy, value) pair produced by one of the static factory
/// methods and passed to <see cref="IWebDriver.FindElement(By)"/> /
/// <see cref="IWebDriver.FindElements(By)"/>. Mirrors Selenium 4.x's
/// <c>By.Id("x")</c> grammar. The strategy strings are exactly what the shared
/// engine's <c>by_locator</c> accepts; ClassName uses the W3C-canonical
/// <c>"class name"</c> form.
/// </summary>
public class By
{
    // Protected so the relative-locator support (RelativeBy) can extend By and
    // flow through the ordinary FindElement(By) path, as in mainstream Selenium.
    protected By(string strategy, string value)
    {
        Strategy = strategy;
        Value = value;
    }

    /// <summary>The engine strategy string (e.g. <c>"css selector"</c>).</summary>
    public string Strategy { get; }

    /// <summary>The raw selector value.</summary>
    public string Value { get; }

    public static By Id(string value) => new("id", value);

    public static By Name(string value) => new("name", value);

    public static By ClassName(string value) => new("class name", value);

    public static By CssSelector(string value) => new("css selector", value);

    public static By TagName(string value) => new("tag name", value);

    public static By LinkText(string value) => new("link text", value);

    public static By PartialLinkText(string value) => new("partial link text", value);

    // Mainstream OpenQA.Selenium spells it XPath (capital P); a script doing
    // By.XPath(...) must compile. Xpath kept as an alias for existing callers.
    public static By XPath(string value) => new("xpath", value);

    [Obsolete("Use By.XPath (mainstream capitalization).")]
    public static By Xpath(string value) => XPath(value);

    // --- desktop / native strategies (WinAppDriver, Appium) ----------------
    // Against a native driver the engine does NOT rewrite Id/Name/ClassName to
    // CSS: they are the driver's own UIA AutomationId / Name / ClassName, and a
    // native driver has no CSS engine. So the factories above keep working on
    // desktop; these add the strategies that only exist there. Values match
    // Appium's AppiumBy.

    public static By AccessibilityId(string value) => new("accessibility id", value);

    public static By AndroidUIAutomator(string value) => new("androidUIAutomator", value);

    public static By AndroidViewTag(string value) => new("androidViewTag", value);

    public static By AndroidDataMatcher(string value) => new("androidDataMatcher", value);

    public static By AndroidViewMatcher(string value) => new("androidViewMatcher", value);

    public static By IOSPredicate(string value) => new("iOSPredicateString", value);

    public static By IOSClassChain(string value) => new("iOSClassChain", value);

    public static By Image(string value) => new("image", value);

    public static By Custom(string value) => new("custom", value);

    public override string ToString() => $"By.{Strategy}: {Value}";
}

File: java/src/main/java/org/openqa/selenium/By.java
package org.openqa.selenium;

/**
 * A locator: a (strategy, value) pair produced by one of the static factory
 * methods and passed to {@code findElement(By)} / {@code findElements(By)}.
 * Mirrors Selenium 4.x's {@code By.id("x")} grammar. The strategy strings are
 * exactly what the shared engine's {@code by_locator} accepts; className uses
 * the W3C-canonical {@code "class name"} form.
 */
public class By {

    private final String strategy;
    private final String value;

    /**
     * Package/subclass constructor. Kept accessible so the relative-locator
     * support ({@code support.locators.RelativeLocator.RelativeBy}) can extend
     * {@code By} and flow through the ordinary {@code findElement(By)} path, as
     * in mainstream Selenium.
     */
    protected By(String strategy, String value) {
        this.strategy = strategy;
        this.value = value;
    }

    /** The engine strategy string (e.g. {@code "css selector"}). */
    String strategy() {
        return strategy;
    }

    /** The raw selector value. */
    String value() {
        return value;
    }

    public static By id(String value) {
        return new By("id", value);
    }

    public static By name(String value) {
        return new By("name", value);
    }

    public static By className(String value) {
        return new By("class name", value);
    }

    public static By cssSelector(String value) {
        return new By("css selector", value);
    }

    public static By tagName(String value) {
        return new By("tag name", value);
    }

    public static By linkText(String value) {
        return new By("link text", value);
    }

    public static By partialLinkText(String value) {
        return new By("partial link text", value);
    }

    public static By xpath(String value) {
        return new By("xpath", value);
    }

    // --- desktop / native strategies (WinAppDriver, Appium) ----------------
    // Against a native driver the engine does NOT rewrite id/name/className to
    // CSS: they are the driver's own UIA AutomationId / Name / ClassName, and a
    // native driver has no CSS engine. So the factories above keep working on
    // desktop; these add the strategies that only exist there. Values match
    // Appium's AppiumBy, so a script written against an Appium client reads the
    // same here.

    public static By accessibilityId(String value) {
        return new By("accessibility id", value);
    }

    public static By androidUIAutomator(String value) {
        return new By("androidUIAutomator", value);
    }

    public static By androidViewTag(String value) {
        return new By("androidViewTag", value);
    }

    public static By androidDataMatcher(String value) {
        return new By("androidDataMatcher", value);
    }

    public static By androidViewMatcher(String value) {
        return new By("androidViewMatcher", value);
    }

    public static By iOSPredicate(String value) {
        return new By("iOSPredicateString", value);
    }

    public static By iOSClassChain(String value) {
        return new By("iOSClassChain", value);
    }

    public static By image(String value) {
        return new By("image", value);
    }

    public static By custom(String value) {
        return new By("custom", value);
    }

    /**
     * Find the first element in {@code context} matching this locator (upstream
     * {@code by.findElement(driver)} form). Delegates to
     * {@link SearchContext#findElement(By)}.
     */
    public WebElement findElement(SearchContext context) {
        return context.findElement(this);
    }

    /**
     * Find all elements in {@code context} matching this locator (upstream
     * {@code by.findElements(driver)} form).
     */
    public java.util.List<WebElement> findElements(SearchContext context) {
        return context.findElements(this);
    }

    @Override
    public String toString() {
        return "By." + strategy + ": " + value;
    }

    @Override
    public boolean equals(Object o) {
        return (o instanceof By) && this.toString().equals(o.toString());
    }

    @Override
    public int hashCode() {
        return toString().hashCode();
    }
}
